doc/maxwell.8

.\" Copyright (c) <year>, <copyright holder>
.\"
.\" %%%LICENSE_START(GPLv2+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; version 2.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END

.TH MAXWELL 8 2011-09-30 Linux 

.SH NAME
Maxwell \- Daemon for gathering entropy from a timer and feeding it to
.BR random (4).

.SH SYNOPSIS.
.ll +8
.B maxwell
.RB [ " \-dfgmtxy " ]
.B maxwell
.RB [ " \-<digit> " ]

.B maxwell [\-c 
.I number
]
[\-p 
.I number
]
[\-s 
.I number
]
[\-w 
.I number
]

.SH DESCRIPTION

.I maxwell
collects randomness from the small variations in a system
timer, distills it into a concentrated form, and sends it to
.BR random (4).

Non-root users can run the program only with the -t (test)
option, which sends test data to standard output.

The amount of output varies with the parameters chosen, but is
generally a few K bits per second. The quality
is intended to be very high. Both volume and quality should be
adequate for most applications.

There are a number of other ways to feed entropy to
.BR random (4).
The advantage of
.I maxwell
is that it is small, simple
and only minimally hardware-dependent. The other methods
also have advantages, and in many cases one of them will
be preferable to this one.
 
For fuller discussion, see the design paper listed in
the references.

.SH OPTIONS

Three options with numeric arguments allow changing the parameters
with which the program runs. These may be omitted, in which case
default values are used.

The numeric arguments are required, and must be in the range 0-999.
.TP 
.BI \-c \ n
Claim, how much entropy do we tell random(4) we
are giving it per 32-bit output? Legal values
are 0 to 32. Anything > 32 sets it to 32.

Default is 30.

.TP 
.BI \-d \ n
Delay, microseconds between timer samples.

Default is 97.

.TP 
.BI \-p \ n
Paranoia, affects how much looping is done  before
each output. Number of loops is 2p + 3. Each loop
collects sixteen timer samples for entropy. 

Default is p = 0 which gives three loops. 

.TP 
.BI \-s \ n
Stop after n kilobits of output.

Default is for the program to continue indefinitely.

.PP
The above options may be combined at will.

Additional options are provided for convenience. They are
a handy way to get a useful combination of options. These
options
.I must
be used alone; they may not be combined either
with each other or with the options listed above.

Two are intended for use at startup, to quickly fill
up the entropy pool. They use smaller delay values,
so they are faster. The claim less entropy so that
.BR random (4)
will accept their output more quickly. Both stop
after a fixed amount of output, but they use
different amounts.

.TP
.I -f
equivalent to	-d 41 -c 16 -s 8
TP
.I -g
equivalent to	-d 43 -c 16 -s 16
TP
.I -G
equivalent to   -g   but not daemonizing

.PP
Three are intended to let a cautious system administrator
run a more conservative version of the daemon. They use
higher paranoia settings.
TP
.I -x
equivalent to	-d 103 -p 2
TP
.I -y
equivalent to	-w 107 -p 3
TP
.I -z
equivalent to	-w 109 -p 4

.PP
Each of -f, -g, -G, -x, -y, -z, and the default with
no options uses a different prime number for the wait.

Other options are provided for testing.

Two of these use
default parameters, except that they send output to standard
out instead of to
.I /dev/random.

.TP
.I -m
Produce one megabit of test output.

.TP
.I -t
Produce test output indefinitely.

.TP
.I -<digit>
This runs the given number of loops, where any other combination of options runs
at least three. It also omits the extra mixing normally done outside
the loop. It stops after a megabit of output.

.PP
If
.I maxwell -1 | ent
shows >= three bits of entropy per byte, then
.I maxwell
is safe. That is 12 bits per loop and
.I maxwell
uses at least three loops per output.

.SH BUGS

The exact entropy generated by a timer call cannot be known, so it
is impossible to know exactly how much entropy the program feeds to
.BR random (4).

The program uses a loop to collect entropy. Each pass makes 16 timer
calls and gets 16 samples modulo 31. With the default settings, there
are three passes for each 32-bit chunk of output. This is safe if 16
timer calls give 11 or more bits of entropy.

There is considerable evidence that available entropy is much more
than this minium requirement. See the design paper.

.SH TESTING

In addition to the -t option described above, there is s test
program called test.sam.c in the source directory. It is designed
to have its output piped into ent(1), an entropy testing program.

This tests at a lower level than maxwell -t; it samples data in
the same way maxwell does, but does no mixing at all, just packs
each sample into a byte and sends it out.

See the design paper and comments in the test program
source for details.

On any unusual system, these tests should be run
before
.BR maxwell ()
is deployed.

.SH SEE ALSO

.BR random (4)
.BR ent (1)

.SH References

RFC 4086
.I Randomness Requirements for Security
is the standard reference on randomness issues.
http://www.ietf.org/rfc/rfc4086.txt

.I Analysis of inherent randomness of the Linux kernel
by Nicholas McGuire, Peter Okech & Georg Schiesser
discusses timer entropy and gives measurements.
http://lwn.net/images/conf/rtlws11/papers/paper.37.html

.I The maxwell(8) random number generator
by Sandy Harris documents the design rationale for
this program.

That paper is included in PDF format in the
source directory for the daemon. If you need
source, under a CC-by-SA license, email me.  


src/maxwell.c
/*
	A daemon to get entropy from timer randomness
	and feed it to random(4)

	It borrows some code from Folkert van Heusden's
	timer entropy daemon http://www.vanheusden.com/te/
	but the two programs are quite different.

	License is GPL v2, the same as the earlier code.

	If anyone needs it under another Open Source
	license, they can contact me.

	Sandy Harris, sandyinchina@gmail.com


	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; version 2.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
	MA  02110-1301, USA.

*/

#include "maxwell.h"
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <syslog.h>

#include <linux/random.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

int good_opt(char *) ;
int good_num(char *) ;

void usage() ;

/*
	minimum number of loops
*/
#define MIN_LOOPS 3

/*
	global variables
	set in main()
	used by log_error() and error_exit()
*/
char *prog_name ;
int user, demon ;

/*
	primes taken from
	http://primes.utm.edu/lists/small/1000.txt

	used as microsecond delay values
*/
unsigned primes[] = {41,43,97,101,103,107,109,113,127} ;

/*
	constants borrowed from SHA-1
	reference used, RFC 3174

	used as initialisation constants
	for the entropy-collecting variable
*/
u32 sha_c[] = {0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476, 0xC3D2E1F0} ;

/*
	constant for multiplications

	11, 37 and 71 are 3, 5 and 7 mod 8
	so they all give some mixing in the low bits
	and they each do it differently
*/
#define MUL (71+(37<<12)+(11<<19))

int main( int argc, char **argv)
{
	unsigned a, b, p, delay, limit, loops ;
	int i, j, out, output, foreground, ret, claim, mul, mix ;
	char *u, *v ;

	/*
		defaults
		might be overidden by options
	*/
	p = 0 ;			// no paranioa, three loops
	delay = primes[2] ;
	limit = 0 ;		// loop forever
	claim = 30 ;		// entropy claim per output
	/*
		defaults
		always apply except in testing
	*/
	demon = 1 ;
	mix = 1 ;
	mul = MUL ;
	/*
		default unless overriden by demon=1
		output to standard out
	*/
	output = 1 ;
	foreground = 0 ;

	/* argument processing */
	prog_name = *argv ;
	switch(argc)	{
		// no options, use defaults
		case 1:
			break ;
		// one, to four name/value pairs
		// no check for repeated options
		// so if same option twice, 2nd is used
		case 9:
		case 7:
		case 5:
		case 3:
			for( argc--, argv++ ; argc ; argc -= 2, argv += 2)	{
				u = argv[0] ;
				v = argv[1] ;
				if( good_opt(u) && good_num(v) )
					a = atoi(v) ;
				else	usage() ;
				if( !strcmp(u,"-p") )
					p = a ;
				else if( !strcmp(u,"-d") )
					delay = a ;
				else if( !strcmp(u,"-s") )
					limit = a ;
				else if( !strcmp(u,"-c") )
					if( a <=32 )
						claim = a ;
					else
						claim = 32 ;
				else
					usage() ;
			}
			break ;
		// one of the standalone options
		case 2:
			// get the option value
			i = good_opt(argv[1]) ;
			// fprintf(stderr, "option -> %c\n", i) ;

			// digit -> loops with no mixing
			if( isdigit(i) )	{
				demon = 0 ;
				limit = 1024 ;
				mix = 0 ;
				mul = 1 ;
				loops = (i -'0') ;
			}
			else switch(i)	{
				// faster variants
				// reduce claim so random(4) accepts
				// inputs faster
				case 'f' :
					delay = primes[0] ;
					claim = 16 ;
					limit = 8 ;
					break ;
				case 'g' :
					delay = primes[1] ;
					claim = 16 ;
					limit = 16 ;
					break ;
				// as -g but don't daemonize
				case 'G' :
					delay = primes[1] ;
					claim = 16 ;
					limit = 16 ;
					foreground = 1;
					break ;
				// for the paranoids
				// -x, -y -z do more loops
				case 'x' :
					delay = primes[3] ;
					p = 2 ;
					break ;
				case 'y' :
					delay = primes[4] ;
					p = 3 ;
					break ;
				case 'z' :
					delay = primes[5] ;
					p = 4 ;
					break ;
				// options for testing
				case 'm' :
					// one megabit
					limit = 1024 ;
					demon = 0 ;
					break ;
				case 't' :
					// zero delay, faster test
					delay = 0 ;
					demon = 0 ;
					break ;
				default:
					usage() ;
					break ;
			}
			break ;
		default:
			usage() ;
			break ;
	}

	// documented in getuid(1) as "always successful"
	// so do not bother checking for failure
	user = geteuid() ;

	/*
		set up various things
	*/

	// not testing and not root: fail
	if( demon && (user != 0) )	{
		fprintf(stderr, "%s: must be root to run maxwell(8)\n", prog_name) ;
		fprintf(stderr, "%s: updates kernel entropy count, so root only\n", prog_name) ;
		fprintf(stderr, "%s: exception is 'maxwell -[t|m|0-9]' for testing\n", prog_name) ;
		exit(1) ;
	}

	// not testing and root
	if( demon )	{
		if( (output=open("/dev/random", O_WRONLY)) == -1)
			error_exit("failed to open /dev/random") ;
		if( !foreground && daemon(-1, -1) == -1)
			error_exit("failed to become daemon process") ;
		openlog( prog_name, (LOG_CONS|LOG_PID), LOG_DAEMON) ;
	}

	// calculate loops per output
		loops = MIN_LOOPS + 2*p ;

	// limit is given in K bits
	// but we need the number of 32-bit chunks
	if( limit )
		limit *= 32 ;

	message( "maxwell(8) v 1.2" );

	for( out = 0 ; (limit == 0) || (out < limit) ; out++) {
		if( mix )
			// initialise with a somewhat random constant
			a = sha_c[t5()] ;
		else
			a = 0 ;
		// main loop; sample & mix
		for( i = 0 ; i < loops ; i++)	{
			// get 16 samples for entropy
			for( j = 0 ; j < 16 ; j++ )	{
				usleep(delay) ;
				// mix in a sample
				// multiplication spreads bits out
				a += ( t31() * mul ) ;
				// rotate left by two bits
				b = a >> 30 ;
				a = (a << 2) | b ;
			}
			if( loops > 1 )	{
				// rotate by one bit between loops
				b = a >> 31 ;
				a = (a << 1) | b ;
			}
		}
		// mix thoroughly
		if( mix )
			a = qht(a) ;
		// output 32 bits
		if( (ret = write(output, &a, 4)) != 4)
			error_exit("write to /dev/random failed") ;
		// update entropy estimate
		if( demon )
			// a failure here is logged
			// but does not stop the program
			// it does no real harm
			if( (ioctl(output, RNDADDTOENTCNT, claim)) == -1)
				message("ioctl() failed") ;
	}
	message( "shutting down" );
	exit(0) ;
}

void error_exit(const char *reason)
{
	if( demon )		// running as root
		syslog(LOG_ERR, "%s", reason) ;
	else
		fprintf(stderr, "%s: %s\n", prog_name, reason) ;
	exit(-1) ;
}

void message(const char *reason)
{
	if( demon )
		syslog(LOG_INFO, "%s", reason) ;
	// else
	//   do nothing
}

void usage()
{
	fprintf(stderr, "usage: %s [-p <number>] [-d <number>] [-s <number>] [-c number]\n", prog_name) ;
	fprintf(stderr, "usage: %s [-f|-g|-G|-m|-t|-x|-y|-z|-<digit>]\n", prog_name) ;
	exit(-1) ;
}

/*
	is this a legal single character option?

	no program-specific checks
	just is it '-' followed by a single letter or digit
	if so, return the letter or digit
*/
int good_opt(char *p)
{
	if( (p[0] == '-') && isalnum(p[1]) && (p[2] == '\0') )
		return(p[1]) ;
	else
		return(0) ;
}

/*
	allows numbers 0 - 999, no + - or decimal point
*/
int good_num(char *p)
{
	int i ;
	for( i = 0 ; *p ; p++, i++ )	{
		if( isdigit(*p)	)
			// do nothing
			 ;
		else
			return(0) ;
	}
	return( (i>0) && (i<=3) ) ;
}

